src/commissaire_service/clusterexec/__init__.py

```diff
@@ -0,0 +1,329 @@
+                except Exception as error:
+                    # If there was a failure, set the end_status and break.
+                    end_status = 'failed'
+                    self.logger.error(
+                        'Clusterexec {} for {} failed: {}: {}'.format(
+                            command_name, host.address, type(error), error))
+                    break
+
+            # Set the finished hosts.
+            finished_hosts = getattr(model_instance, finished_hosts_key)
+            finished_hosts.append(host.address)
+            try:
+                index = model_instance.in_process.index(host.address)
+                model_instance.in_process.pop(index)
+            except ValueError:
+                self.logger.warn(
+                    'Host {} was not in_process for {} {}'.format(
+                        host.address, command_name, cluster_name))
+            try:
+                params = {
+                    'model_type_name': model_instance.__class__.__name__,
+                    'model_json_data': model_instance.to_json()
+                }
+                self.request('storage.save', params=params)
+            except Exception as error:
+                self.logger.error(
+                    'Unable to save cluster state for "{}" clusterexec '
+                    'due to {}: {}'.format(cluster_name, type(error), error))
+                raise error
+
+            self.logger.info(
+                'Finished executing {} for {} in {}'.format(
+                    command_name, host.address, cluster_name))
+
+        # Final set of command result.
+
+        model_instance.finished_at = datetime.datetime.utcnow().isoformat()
+        model_instance.status = end_status
+
+        self.logger.info(
+            'Cluster {} final {} status: {}'.format(
+                cluster_name, command_name, model_instance.to_json()))
+
+        try:
+            params = {
+                'model_type_name': model_instance.__class__.__name__,
+                'model_json_data': model_instance.to_json()
+            }
+            self.request('storage.save', params=params)
+        except Exception as error:
+            self.logger.error(
+                'Unable to save final state for "{}" clusterexec '
+                'due to {}: {}'.format(cluster_name, type(error), error))
+
+    def on_upgrade(self, message, cluster_name):
+        """
+        Executes an upgrade command on hosts across a cluster.
+
+        :param message: A message instance
+        :type message: kombu.message.Message
+        :param cluster_name: The name of a cluster
+        :type cluster_name: str
+        """
+        self.logger.info(
+            'Received message: Upgrade cluster "{}"'.format(cluster_name))
+        command_args = ('upgrade',)
+        model_instance = ClusterUpgrade.new(
+            name=cluster_name,
+            status='in_process',
+            started_at=datetime.datetime.utcnow().isoformat(),
```

Comment:
nit: ``commissaire.util.date.formatted_dt`` could be used here since it's been merged.